src/shogun/machine/Composite.cpp

```diff
@@ -0,0 +1,34 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Yuhui Liu
+ */
+#include <memory>
+
+#include <shogun/machine/Composite.h>
+
+using namespace shogun;
+
+
+ std::shared_ptr<Composite> Composite::with(std::shared_ptr<Machine> machine){
+            m_machines.push_back(machine);
+            return shared_from_this()->as<Composite>();
+    }
+std::shared_ptr<Composite> Composite::then(std::shared_ptr<CombinationRule> rule){
+            m_rule = rule;
+            return shared_from_this()->as<Composite>();
+}
+std::shared_ptr<BaggingMachine> Composite::fit(std::shared_ptr<Features> data, std::shared_ptr<Labels> y){
+    auto curr_data = data;
+    for(auto && tran: m_trans){
+        curr_data = tran->transform(curr_data);
+    }
+    auto bagging = std::make_shared<BaggingMachine>(); 
+    for(auto&& machine : m_machines){
+        bagging->set_labels(y);
+        bagging->set_machine(machine);
+        bagging->train(data);
```

Comment:
maybe you could put this in a std::thread, so that the bagging is done in parallel

Reply:
btw what you would need here is a BaggingMachine that can handle several different Machine objects. This is just bootstrapping from the features and applying the same Machine N times, I think? @karlnapf I don't think we have an actual EnsembleMachine for this?

Reply:
I dont think we have, but I am not sure. Think @vigsterkr wrote this stuff

Reply:
be mindful that the train method is not (yet) const. (It should be, one of the things @LiuYuHui could look into after the refactor)

Reply:
actually, it's not applying the same Machine N times, the `train_machine` method of `BaggingMachine` train `m_machine`, then put the trained `machine` to the `m_bags`, so i have to repeatedly set `m_machine` and `train`.
https://github.com/shogun-toolbox/shogun/blob/e29377db4f95b90b52425de084bf6d6ac39d2aba/src/shogun/machine/BaggingMachine.cpp#L159-L219

Reply:
but if you scroll up you see that m_bags is cleared before entering the loop

Reply:
https://github.com/shogun-toolbox/shogun/blob/e29377db4f95b90b52425de084bf6d6ac39d2aba/src/shogun/machine/BaggingMachine.cpp#L159-L219

Reply:
oh, i see, so `BaggingMachine` can only train the same Machine N times? that's weird

Reply:
Not really, see on [wikipedia](https://en.wikipedia.org/wiki/Bootstrap_aggregating) ("Bagging is a special case of the model averaging approach"). Model averaging (or ensemble learning) is the more general algorithm, which is what you need here. But it seems like we don't have it in shogun.

Reply:
thanks, so it seems like that i should add the `EnsembleMachine` class.

Reply:
Yes I think so. it shouldn't be too difficult. You can just have a loop over a vector of Machine in the train machine class function and put a `#pragma parallel` over the loop. You could actually test and see if it would be better to write it with `std::thread`, but that is probably more for later on